marlowe-playground-client/src/Halogen/CurrencyInput.purs

```diff
@@ -0,0 +1,115 @@
+module Halogen.CurrencyInput where
+
+import Prelude hiding (div)
+import Data.Array (filter)
+import Data.BigInteger (BigInteger, fromString)
+import Data.Maybe (Maybe(..), maybe)
+import Data.Monoid (guard)
+import Data.String (trim)
+import Data.String.CodeUnits (fromCharArray, toCharArray)
+import Data.Traversable (for)
+import Effect (Effect)
+import Effect.Uncurried (EffectFn1, EffectFn2, runEffectFn1, runEffectFn2)
+import Effect.Unsafe (unsafePerformEffect)
+import Halogen.Css (classNames)
+import Halogen.HTML (HTML, div, input, text)
+import Halogen.HTML.Events (onChange, onFocus, onInput, onKeyDown)
+import Halogen.HTML.Properties (InputType(..), type_, value)
+import Pretty (showBigIntegerAsCurrency)
+import Web.Event.Event (currentTarget)
+import Web.Event.Internal.Types (EventTarget)
+import Web.UIEvent.FocusEvent as FocusEvent
+import Web.UIEvent.KeyboardEvent as KeyboardEvent
+
+foreign import setOldValues_ :: EffectFn1 EventTarget Unit
+
+foreign import checkChange_ :: EffectFn1 EventTarget Unit
+
+foreign import formatValue_ :: EffectFn2 EventTarget Int Unit
+
+foreign import formatValueString_ :: EffectFn2 String Int String
+
+foreign import setValue_ :: EffectFn2 EventTarget String Unit
+
+foreign import onChangeHandler_ :: EffectFn2 EventTarget Int String
+
+setOldValues :: EventTarget -> Effect Unit
+setOldValues = runEffectFn1 setOldValues_
+
+checkChange :: EventTarget -> Effect Unit
+checkChange = runEffectFn1 checkChange_
+
+setValue :: EventTarget -> String -> Effect Unit
+setValue = runEffectFn2 setValue_
+
+onChangeHandler :: EventTarget -> Int -> Effect String
+onChangeHandler = runEffectFn2 onChangeHandler_
+
+currencyInput :: forall p a. Array String -> BigInteger -> String -> Int -> (Maybe BigInteger -> Maybe a) -> HTML p a
+currencyInput classList number prefix rawNumDecimals onValueChangeHandler =
+  div [ classNames ([ "bg-gray-light", "flex", "items-center", "p-0", "m-0", "border-solid", "border", "rounded-sm", "overflow-hidden", "box-border", "focus-within:ring-1", "focus-within:ring-black" ] <> classList) ]
+    ( ( guard hasPrefix
+          [ div [ classNames [ "flex-none", "px-2", "py-0", "box-border", "self-center" ] ]
+              [ text prefix ]
+          ]
+      )
+        <> [ input
+              [ classNames [ "flex-1", "text-center", "box-border", "self-stretch", "border-0", "outline-none" ]
+              , onFocus
+                  ( \ev ->
+                      maybe Nothing
+                        ( \target ->
+                            unsafePerformEffect
```

Comment:
This is a red flag to have in a component. I think we could refactor this into being a proper halogen component.

Reply:
Could you be more specific about what the problem with this approach is?
I am aware I am kind of hijacking the Halogen control, but that is on purpose, because the default mechanism just doesn't work properly. I considered creating a whole new component, but I think that would be much more problematic, because I think we will have to write a whole lot of boiler plate and there will be many more opportunities for errors.

Reply:
Well... I think any discussion around this would boil down to these questions:
1) Does it makes sense to do this project in PureScript instead of plain JS or TypeScript?
2) If the answer is yes, does it make sense to use Halogen instead of something with better interop with the ecosystem, like the react bindings?

Without getting too much into it, if we have to choose again, I'd still make this project in PS but I probably chose the react bindings. I was really glad that we chose PS when I was making parts of the dashboard and I reutilized a lot of functionality from the playground. For stuff like the tooltip that I'm doing, or this special input, doing a react component would probably have less boilerplate.

In the current state of things, I don't think changing languages or frameworks is worth our time, and at the very least I think is worth making this a proper Halogen component for the sake of consistency. 

On a more technical aspect, I don't even know what type of problems this could potentially bring, which is the reason why I'd like to avoid it. Halogen works with a virtual dom and you are modifying the dom directly. I don't know enough of Halogen's internals to say in which cases this could blow up. You mentioned that having a lot of boilerplate could bring more opportunities for making a mistake, but I'd argue that messing with the internals of a library is more dangerous. Especially when the boilerplate is typed checked and the hack isn't.

To reduce the boilerplate, you can make the same choice that I'm doing with the tooltip (PR coming soon). Most of our components are actually "subcomponents", which is a concept that David and Kris created so we can have all/most of the components sharing a single state and use lenses to reduce the scope when we want to. The subcomponents adds boilerplate for the actions and the render, needing to wrap actions and bubble up and down. This boilerplate is fine for pages, but a little annoying for small widgets like this input (or the tooltip). For that, a proper Halogen component might be best suited. Then, the boilerplate is mostly at the view level, and basically you just need to make sure that the view you are inserting has a signature of `ComponentHTML Action ChildSlots m` instead of the more restrictive `HTML p Action`. If you do that, then the rest of the logic is not more different than having a normal input component. The main difference is that for a normal input component you use the `props` to fire actions, and in here you'd use the new component's output message.

Reply:
I imagine the Halogen component simply wraps the HTML input component and it creates a new one when it redraws, and my modifications are all local to the input component, so I really doubt it is a big deal. Plus you guys will soon be redesigning the playground simulator and then you can do it more properly. In the meanwhile, I think this is much better than what we currently have for user experience

Reply:
The Halogen component does a little bit more, I know that it has an internal reference of the VDOM and it does some diffing with the actual DOM and then tries to do the minimum modifications so that the DOM looks like the VDOM says it should look. At some point while testing, I manually modified the DOM to try something and the changes were overwritten by Halogen... Once again, I think it's dangerous to use a tool in a way that wasn't meant to be used, but it's your call.

Reply:
That is interesting, you are right: it seems the VDOM must store references to the DOM nodes, if you delete them it notices and places them back (or recreates them if they have been garbage collected). But I still think risk is very low in this case, and since it is a temporary solution and it is quite hacky in any case, I don't think it is worth the trouble at the moment, I'll assume the risk. But, of course, you make a good point